### PR TITLE
Configure DNS64 and NAT64 for IPv6 subnets

### LINKED ISSUE #4505: Configure DNS64 and NAT64 for IPv6 subnets
Part of https://github.com/weaveworks/eksctl/issues/4255

**Description**
Background info: https://docs.aws.amazon.com/vpc/latest/userguide/vpc-nat-gateway.html#nat-gateway-nat64-dns64

DNS64 and NAT64 must be configured on all IPv6 subnets to enable communication with IPv4-only services.

**1. Enable DNS64**
Add functionality so that, when creating IPv6 subnets, all subnets  (except for fully private ones) are created with the `enableDNS64` param.

**2. Add route for NAT64**
In the current [worfklow](https://github.com/weaveworks/eksctl/blob/ipv6/pkg/cfn/builder/vpc_ipv6.go), we already create the `0.0.0.0/0` <-> NATGW route and the `::0` <-> NATGW route for all subnets. 

We only need to create a route between the NATGW and `64:ff9b::/96` in the IPv6 CIDR. 

**3. Update SG**
Modify the cluster SG and NACLs to allow egress and ingress traffic to `64:ff9b::/96` traffic.

**Scope**
- Modify subnets to enable DNS64
- Create an additional `EC2::Route` for NAT64 to translate IPv4 traffic
- Update SG role 

**Acceptance Criteria**
As an eksctl user with an IPv6 cluster,
I should be able to see that DNS64 is enabled on all my subnets,
And I should be able to see that NAT64 translates IPv4 traffic,
And I should be able to make an IPv4 request to a service running on my IPv6 cluster.

### PR DESCRIPTION
**Pull request by aclevername**
## Description

Close #4505

### Implementation Details
We followed [this](https://docs.aws.amazon.com/vpc/latest/userguide/vpc-nat-gateway.html#nat-gateway-nat64-dns64) AWS guide on how to enable DNS64 and NAT64 for IPv6-only subnets. 

#### 1. Enabling DNS64 on private subnets

DNS64 and NAT64 are intended for subnets that have IPv6-only workloads and use a NAT Gateway. In our case, this means private subnets configured to use IPv6. 

DNS64 is enabled on all private subnets as a post-creation task. We will need to create this as part of the subnet creation process once CFN has support for `DNS64`. Ticket that tracks this issue: https://github.com/weaveworks/eksctl/issues/4540

#### 2. Adding route pointing to NAT64

The IPv4 route `0.0.0.0/0` to the NAT Gateway and IPv6 route `::0` to the EOIGW mentioned in the guide are already created in the default IPv6 workflow, so they did not need to be created as part of this PR. 

The only route that needed to be added is a route for IPv6 traffic prefixed by the [DNS64 Well Know Prefix `64:ff9b::/96`](https://datatracker.ietf.org/doc/html/rfc6147) pointing to the NAT Gateway.

### Testing

Before implementing the DNS64 and NAT64 changes, we created an IPv6 cluster and then created a nodegroup on one of its private subnets. We used one of the public subnets to SSH onto the private subnet. From there, we checked whether we can query to a IPv4-only website over IPv6:
```
curl -6 https://ipv4.jamieweb.net/
curl: (7) Couldn't connect to server
```
This failed.

We implemented the DNS64/NAT64 changes and followed these steps again:
```
curl -6 https://ipv4.jamieweb.net/
<html>                                                                                                                                                          
<head><title>IPv4 Test Page</title></head>                                                                                                                    
<body>  
<h2>IPv4 Test Page</h2> 
This page is only accessible over IPv4 (139.162.222.67).      
...
```
Which was successful! This means services are successfully able to query an IPv4 IP address over IPv6 from a private subnet. :)

### CODE REVIEW

**Comment by aclevername** [comment]
> Jesus Christ this gave me a little heart attack. `+3,948 −20`.

mb :smile: 

WIP PR anyway as it doesn't work currently so don't review it!

**Comment by nikimanoledaki** [comment]
Tested this then filled in the description with implementation details and how this was tested! 👍

**Comment by Skarlso** [comment]
@aclevername @nikimanoledaki is this now ready to be reviewed?

**Comment by nikimanoledaki** [comment]
It is from my side, just waiting for @aclevername to confirm as well. :)

**Comment by aclevername** [comment]
Go for it @Skarlso!

File: go.mod [hunk 1]
```diff
@@ -398,8 +398,8 @@ require (
 
 replace (
 	// TODO: once aws-sdk-go is updated with ipv6 code remove the below line
-	github.com/aws/aws-sdk-go => github.com/weaveworks/aws-sdk-go v0.0.0-20211026093156-d6e6822f58db
-	// Used to pin the k8s library versions regardless of what other dependencies enforce
+	github.com/aws/aws-sdk-go => github.com/weaveworks/aws-sdk-go v0.0.0-20211208164730-c076f913e1de
+	// Used to pin the k8s library versions regardless of what other dependencies enforc
```
Comment (cPu1) [inline go.mod:402]:
```suggestion
	// Used to pin the k8s library versions regardless of what other dependencies enforce
```

File: pkg/cfn/builder/vpc.go [hunk 2]
```diff
@@ -30,8 +31,10 @@ const (
 	PrivateRouteTableAssociation = "RouteTableAssociationPrivate"
 	PubSubRouteKey               = "PublicSubnetDefaultRoute"
 	PubSubIPv6RouteKey           = "PublicSubnetIPv6DefaultRoute"
+	PubSubDNS64IPv6RouteKey      = "PublicSubnetDNS64IPv6DefaultRoute"
```
Comment (cPu1) [inline pkg/cfn/builder/vpc.go:34]: This is unused. Were you planning to use it somewhere?

Comment (nikimanoledaki) [reply pkg/cfn/builder/vpc.go:34]: Good spot, that can be removed!

**Review by cPu1 — APPROVED** [review]
LGTM!

**Review by Skarlso — APPROVED** [review]

**Comment by Skarlso** [comment]
Well done! :)

**Pull request merged**